### ISSUE TITLE
New lint rule to inform variable names for imported images and audio

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/chipper#1060]
From https://github.com/phetsims/circuit-construction-kit-common/issues/702, and https://github.com/phetsims/circuit-construction-kit-common/issues/702#issuecomment-872475640.

We would like a lint rule that enforces that all variables for imported images end in "Image" and all variables for imported sounds end it "Sound". 

I am envisioning a single lint rule for this. Using https://astexplorer.net/, here is the AST that will inform how we implement the rule for import statements

<details>

```
{
  "type": "Program",
  "body": [
    {
      "type": "ImportDeclaration",
      "specifiers": [
        {
          "type": "ImportDefaultSpecifier",
          "local": {
            "type": "Identifier",
            "name": "relativePathSound",
            "range": [
              7,
              24
            ]
          },
          "range": [
            7,
            24
          ]
        }
      ],
      "source": {
        "type": "Literal",
        "value": "../../../sounds/relative-path-sound_mp3.js",
        "raw": "'../../../sounds/relative-path-sound_mp3.js'",
        "range": [
          30,
          74
        ]
      },
      "range": [
        0,
        75
      ]
    }
  ],
  "sourceType": "module",
  "range": [
    0,
    75
  ]
}
```
</details>

**Comment by jessegreenberg** [comment]
In the above commits I implemented this lint rule. But then realized that a lot of Image Node variables use the "Image" suffix as well, and renaming the import variable caused a name collision. So in order to use this convention for the import they will need to be renamed to have a suffix like "ImageNode". That could be fine, but that would also mean updating tandem names for instrumented images.

Many look like
```
const myImage = new Image(... {
  ...
  tandem: options.tandem.createTandem( 'myImage' )
} );
```

To check in next developer meeting:
- Do we still want the variable for image imports to end in Image?
- Is it important that all PhET-iO instrumented Images have a similar tandem name suffix?
- If yes to both, Image nodes will need a variable and tandem name that is different from the import, that ends in something like "ImageNode".
- There are ~450 Image imports that currently don't follow this rule, will take some time to change them all. An auto fix doesn't seem workable here because lots of the existing variables are named things like "myImg" or "myImage1", and just adding "Image" to the end would be sloppy.

**Comment by samreid** [comment]
From today's meeting, we are considering a different suffix for the imported images and mipmaps.  We would like to turn on the rule so it isn't "wild west/no pattern".

@pixelzoom proposed: matching the variable name to the filename, converting dash to underscore. `import level_3_icon_png from '../../../mipmaps/level-3-icon_png.js';`

**Comment by pixelzoom** [comment]
In the above commits, I converted many of my sims to use strict camelcase for image filenames.  The modulified names and imports then looks like this:

```js
import dropperIcon_png from '../../../images/dropperIcon_png.js';
import shakerIcon_png from '../../../images/shakerIcon_png.js';
```

I like this convention a lot, and I'm going to convert the rest of my sims.

The advantage to converting the file names is theres no need to convert dashes to underscores, and the import names and file names are therefore identical.  If it makes writing the lint rules any easier, I'd vote to change all sims to use this pattern.

**Comment by pixelzoom** [comment]
Image file names in all of my sims have been converted to camelcase, _EXCEPT_ graphing-quadratics https://github.com/phetsims/graphing-quadratics/issues/156 and molecule-polarity https://github.com/phetsims/molecule-polarity/issues/131.  Those 2 sims are in RC testing, and I don't want to complicate potential cherry-picks by changing master.  So I'll address them when they have been published.

@jessegreenberg It occurred to be that if all sims used this convention, then it's the same convention as source-code files.  So could we use the same lint rule?  I also noticed that something like this:

```js
import SomethingBogus from '../../../../scenery/js/nodes/Image.js';
```

... does not result in a lint error. Should it?

**Comment by pixelzoom** [comment]
Proposal:

(1) Change `grunt modulify` to change '-' to '_' in the output file name.
(2) Run `grunt modulify` on all repos to remove '-' in file names. Commit new files, remove old files that were named with '-'.
(3) Fix `import` statements file names that were changed in (3). Identify what needs to be fixed by searching for "_png.js" etc., or by running sims to see which ones now have a bad import.
(4) See if there's a standard lint rule for "import name must match file name".  If there is, enabled it. If not, write one. 
(5) Run `lint everything`.  Fix (or auto-fix) all import names that do not match the file name.

**Comment by pixelzoom** [comment]
In the above commits, image file names and imports were standardized in molecule-polarity, vegas, and scenery-phet.

**Comment by pixelzoom** [comment]
We spent so much time discussing this, it's gone nowhere, and I'm seeing so much NEW code that is not using the new convention (including the sim I'm code reviewing today).  And as @jbphet mentioned at last dev meeting, we need a chip-away to get existing sims moved over to this convention.  So I'm going to move this back into the "triage" list in the Developer Meeting project board, with high priority.

**Comment by pixelzoom** [comment]
> What is the rule for sounds?

We decided that this convention and rule should apply to all "media file" imports. From the meeting notes in https://github.com/phetsims/chipper/issues/1060#issuecomment-906688485:

> (1) Change `grunt modulify` to change '-' to '_' in the output file name.
> (1.5) UPDATE FROM DEV MEETING: change media filenames as much as time permits form '-' to '_'.

**Comment by samreid** [comment]
Sounds good, should I delete the `import-resource-variable-suffix` rule?

**Comment by jessegreenberg** [comment]
Yes that can be deleted, it is going to be replaced by a different lint rule in https://github.com/phetsims/chipper/issues/1079.

**Comment by samreid** [comment]
Thanks, I deleted it.

**Comment by zepumph** [comment]
I believe the main point for discussion at dev meeting is how to prioritize this quite manual work of renaming files, assets, and imports.

**Comment by samreid** [comment]
I added this to status meeting agenda:

SR: For designers and artists, we have a new convention for filenames.  Images, sounds, mipmaps, etc should use camel casing like:
wavePacketHomeScreenIcon.png (GOOD: camel casing)
wave-packet-home-screen-icon.png (BAD: hyphens)
This makes it easier to load into the code
Not strictly necessary for assets/, but consider consistency when adding a source artwork file that corresponds to images/sounds/mipmaps
The convention for sim screenshots has NOT changed.
If you need a non-camel-casing delimiter in rare cases, you can use underscore:
fluteV5_1.mp3 (RARE: underscores)

@jessegreenberg and I discussed it on slack:

Jesse Greenberg  9:43 AM
Hey Sam! For the filename convention item on status agenda, will that still be necessary if we change modulify to convert dashes to underscores?

Sam Reid  9:54 AM
Good question!  I thought the new proposal was to rename the filenames, and I saw a lot of commits from @pixelzoom using camel casing instead of underscores.  Should I remove the agenda item until we can discuss further?

Jesse Greenberg  9:59 AM
Yea, I don't think that is necessary unless it is preferable to have the image and .js files be as similar as possible.

Sam Reid  10:01 AM
To me, it’s preferable to have the filenames as similar as possible.  I’ll move the item to dev meeting though to get team discussion.

Jesse Greenberg  10:02 AM
OK! That works for me. Perhaps we should proceed with that, we already discussed it as a team in https://github.com/phetsims/chipper/issues/1060#issuecomment-906688485.

Sam Reid  10:05 AM
Well, the written consensus is to use _ but not camelcasing, but all of CM commits use camelcasing.  I recommend the camel casing.  I don’t mind waiting one more week to PSA.  What do you recommend?
10:06
Also, in my changes, I preferred camelcasing to underscores, just using them rarely where necessary.

Jesse Greenberg  10:09 AM
I prefer camel casing too. I think we should keep it on the agenda, it is our last status meeting for ~3 weeks with the holidays.

Sam Reid  10:09 AM
OK I’ll restore it.

Jesse Greenberg  10:09 AM
Thanks, sorry for the false alarm!

Sam Reid  10:09 AM
No worries, thanks for discussing!

**Comment by samreid** [comment]
Once https://github.com/phetsims/chipper/issues/1177 is done, we can turn on the new lint rule.  That's all that's left for this issue, the main work will be in the chip-away.

**Comment by jessegreenberg** [comment]
Adding on-hold label for https://github.com/phetsims/chipper/issues/1060#issuecomment-996111085.

**Comment by zepumph** [comment]
Is it purposeful that the lint rule isn't very strict on special characters in the matching?

This passes the lint rule. . . 
```
import selectionArpeggio002_mp3 from '../../../../tambo/sounds/selection-arpeggio-002_mp3.js';
```

**Comment by jessegreenberg** [comment]
I think it should be more strict, from https://github.com/phetsims/chipper/issues/1060#issuecomment-906688485
> Change modulify task (UPDATE DEC 16: We will just fix the image/mipmap/sound filenames with camelcase (or underscores rarely when necessary).

That example passes because of `normalizeFilename` in default-import-match-filename.js. Can we remove "normalizing" in `isCompatible`? This change to the rule makes sense to me, does it work or you @zepumph?

<details>

```patch
Index: eslint/rules/default-import-match-filename.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/eslint/rules/default-import-match-filename.js b/eslint/rules/default-import-match-filename.js
--- a/eslint/rules/default-import-match-filename.js	(revision 50745bb81febca9928d065a597b78f6e3c23328c)
+++ b/eslint/rules/default-import-match-filename.js	(date 1640632099198)
@@ -18,14 +18,6 @@
   return path.basename( filename, path.extname( filename ) );
 }
 
-/**
- * @param {string} filename
- * @returns {string}
- */
-function normalizeFilename( filename ) {
-  return filename.replace( /[-_.]/g, '' ).toLowerCase();
-}
-
 /**
  * Test if local name matches filename.
  * @param {string} localName
@@ -33,11 +25,10 @@
  * @returns {boolean}
  */
 function isCompatible( localName, filename ) {
-  const normalizedLocalName = localName.replace( /_/g, '' ).toLowerCase();
 
   return (
-    normalizedLocalName === normalizeFilename( filename ) ||
-    normalizedLocalName === normalizeFilename( removeExtension( filename ) )
+    localName === filename ||
+    localName === removeExtension( filename )
   );
 }
 
@@ -143,6 +134,7 @@
       if (
         !isCompatible( defaultImportName, filename )
       ) {
+
         context.report( {
           node: defaultImportSpecifier,
           message: `Default import name does not match filename "${filename}".`


```

</details>

**Comment by zepumph** [comment]
Yes thanks. That's totally it. Committed above. I was able to test by noting that after your changes, there are 10 lint errors in molarity, as expectected. I'm also now seeing a total of only 400 lint errors for this across the project. How fun!

I was wondering though it this part is even needed now, can you imagine a usage where the variable could possibly include the file extension given that it would have to keep the period? 

```diff
Index: eslint/rules/default-import-match-filename.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/eslint/rules/default-import-match-filename.js b/eslint/rules/default-import-match-filename.js
--- a/eslint/rules/default-import-match-filename.js	(revision cae7405f9bf16029f5556001b3352fe11b7bafe1)
+++ b/eslint/rules/default-import-match-filename.js	(date 1641232357561)
@@ -25,7 +25,7 @@
  * @returns {boolean}
  */
 function isCompatible( localName, filename ) {
-  return localName === filename || localName === removeExtension( filename );
+  return localName === removeExtension( filename );
 }
 
 /**
```

**Comment by jessegreenberg** [comment]
Thanks! I can't think of why we would need (or could even have) a file extension in the variable name. Ill remove it.

**Comment by jessegreenberg** [comment]
I applied your patch and ran the lint rule and the output looked appropriate. Committed above. Good to close again?

**Comment by zepumph** [comment]
From https://github.com/phetsims/chipper/issues/1060#issuecomment-996111085:

>Once #1177 is done, we can turn on the new lint rule. That's all that's left for this issue, the main work will be in the chip-away.


Marking on hold again.

**Comment by jessegreenberg** [comment]
> Once https://github.com/phetsims/chipper/issues/1177 is done, we can turn on the new lint rule. That's all that's left for this issue, the main work will be in the chip-away.

The lint rule was turned on in that issue, nothing more to do here. Closing.
